x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/stringstats/StringStatsAggregator.java

```diff
@@ -0,0 +1,176 @@
+            minLength = bigArrays.newIntArray(1, false);
+            minLength.fill(0, minLength.size(), Integer.MAX_VALUE);
+            maxLength = bigArrays.newIntArray(1, false);
+            maxLength.fill(0, maxLength.size(), Integer.MIN_VALUE);
+            charOccurrences = new HashMap<>();
+        }
+        this.format = format;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return valuesSource != null && valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final BigArrays bigArrays = context.bigArrays();
+        final SortedBinaryDocValues values = valuesSource.bytesValues(ctx);
+
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                if (bucket >= count.size()) {
+                    final long from = count.size();
+                    final long overSize = BigArrays.overSize(bucket + 1);
+                    count = bigArrays.resize(count, overSize);
+                    totalLength = bigArrays.resize(totalLength, overSize);
+                    minLength = bigArrays.resize(minLength, overSize);
+                    maxLength = bigArrays.resize(maxLength, overSize);
+                    minLength.fill(from, overSize, Integer.MAX_VALUE);
+                    maxLength.fill(from, overSize, Integer.MIN_VALUE);
+                }
+
+                if (values.advanceExact(doc)) {
+                    final int valuesCount = values.docValueCount();
+                    count.increment(bucket, valuesCount);
+
+                    for (int i = 0; i < valuesCount; i++) {
+                        BytesRef value = values.nextValue();
+                        if (value.length > 0) {
+                            String valueStr = value.utf8ToString();
+                            int length = valueStr.length();
+                            totalLength.increment(bucket, length);
+
+                            // Update min/max length for string
+                            int min = Math.min(minLength.get(bucket), length);
+                            int max = Math.max(maxLength.get(bucket), length);
+                            minLength.set(bucket, min);
+                            maxLength.set(bucket, max);
+
+                            // Parse string chars and count occurrences
+                            for (Character c : valueStr.toCharArray()) {
+                                LongArray occ = charOccurrences.get(c);
+                                final long overSize = BigArrays.overSize(bucket + 1);
```

Comment:
It's not a terribly expensive call, but we should be able to move this up and out of the `valuesCount` loop I think?  That way we only calculate the bigarrays size once instead of for each character in each value?

Reply:
When deserializing the map, we use method
```
        charOccurrences = in.<String, Long>readMap(StreamInput::readString, StreamInput::readLong);
```

Not sure if `StreamInput::readString` can work with characters, so I decided to transform the map to <String, Long>